### DISCUSSION TITLE
RuntimeError: No parameter values found

### DISCUSSION THREAD

**Discussion by wissamkafa** [hilo-mpc/hilo-mpc#32] [Q&A] [answered]
Dear @BrunoMorabito 
Hi again.

I am using the learn-MPC example to create a system simulate and replace the MPC with a NN.
I defined   the system equations as follows:

        system.set_dynamical_equations(['x_1',
                                'x_2',
                                'A1*x_1 + B1*u_0 + C1 * Td1',
                                'x_4',
                                'x_5',
                                'A2*x_4 + B2*u_1 + C2*Td2'])

Where A1, B1, C1, A2, B2, C2 are some long math expressions related to the model parameters.
when running the code, I get the error

File "/media/user/.../main_.py", line 162, in <module>
    data_set = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)
  File "/home/user/anaconda3/envs/HILO/lib/python3.9/site-packages/hilo_mpc/modules/dynamic_model/dynamic_model.py", line 4248, in generate_data
    data_generator = DataGenerator(self, **kwargs)
  File "/home/user/anaconda3/envs/HILO/lib/python3.9/site-packages/hilo_mpc/util/data.py", line 673, in __init__
    raise RuntimeError("No parameter values found. Please supply parameter values or set values for the "
RuntimeError: No parameter values found. Please supply parameter values or set values for the parameters of the model before generating data!

And when I print the system I get:
ODEs
         dx_0/dt = x_1
         dx_1/dt = x_2
         dx_2/dt = (((A1*x_1)+(B1*u_0))+(C1*Td1))
         dx_3/dt = x_4
         dx_4/dt = x_5
         dx_5/dt = (((A2*x_4)+(B2*u_1))+(C2*Td2))

which means that those parameters could not take their pre-computed values.
and when I put them explicitly within the equations, it runs with no problems.

I also tried to define a new model like:

def gsys():

        model = Model( name='model_name')
        x = model.set_dynamical_states('x_0', 'x_1', 'x_2', 'x_3', 'x_4', 'x_5')
        u = model.set_inputs('u_0', 'u_1')
        p = model.set_parameters('A1', 'B1', 'C1', 'A2', 'B2', 'C2')

        # Unwrap states
        x_0 = x[0]
        x_1 = x[1]
        x_2 = x[2]
        x_3 = x[3]
        x_4 = x[4]
        x_5 = x[5]

        # Unwrap inputs
        u_0 = u[0]
        u_1 = u[1]

        # Unwrap parameters
        A_1 = p[0]
        B_1 = p[1]
        C_1 = p[2]
        A_2 = p[3]
        B_2 = p[4]
        C_2 = p[5]

        A1 = math_expr
        B1 = math_expr
        C1 = math_expr
        Td1 = const

        A2 = math_expr
        B2 = math_expr
        C2 = math_expr
        Td2 = const
        # ===========================
        dx_0 = x_1
        dx_1 = x_2
        dx_2 = A1*x_1 + B1*u_0 + C1*Td1
        dx_3 = x_4
        dx_4 = x_5
        dx_5 = A2*x_4 + B2*u_1 + C2*Td2
        
        model.set_dynamical_equations([dx_0, dx_1, dx_2, dx_3, dx_4, dx_5])

        return model
       
when printing the system I get:
ODEs
        dx_0/dt = x_1
        dx_1/dt = x_2
        dx_2/dt = ((56069.6*x_1)+(1955.92*u_0))
        dx_3/dt = x_4
        dx_4/dt = x_5
        dx_5/dt = ((-34690.5*x_4)+(1878.31*u_1))

which means the parameters where inserted into the model correctly.

but the previous error remains when trying to generate data in the same way as before.
( data_set = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True))
when reviewing the generate_data function i actually could not figure how this problem is happening.

I also created a dictionary with  those parameters and tried to set them using set_initial_parameter_values(dict), but got the same result.

Any ideas how to fix this or if there is another way to define the system? maybe I am missing something here, and your advise would be very appreciated.

**Comment by brunomorampc** [comment]
HI @wissamkafa can you send me a code that reproduces the error?

**Reply by wissamkafa** [reply]
Yes, I could send the code.
may I have your email?. I don't know if i can upload a file here
my email is: ww.kafa@gmail.com
you may send me a test email and I will send you the file.
or if there is another way, please let me know.

**Comment by wissamkafa** [comment]
For now and while debugging the error, I think it is generated by this piece of code in data.py

       672       if module.n_p > 0 and p0 is None:
       673          raise RuntimeError("No parameter values found. Please supply parameter values or set values for the "
       674                            "parameters of the model before generating data!")
    
But what parameters actually represent?

**Reply by brunomorampc** [reply]
I'd like to see how you are passing the parameters. It looks like they are not correctly passed. I have just sent you an email :)

**Reply by brunomorampc** [reply, marked as answer]
@wissamkafa I sent you an email. You were defining parameters that where never used.
